chunk.go

```diff
@@ -0,0 +1,306 @@
+}
+
+func (lcd *largeChunkDecoder) decode() ([][]*string, error) {
+	if '[' != lcd.nextByteNonWhitespace() {
+		return nil, lcd.mkError("expected chunk to begin with '['")
+	}
+
+	rows := make([][]*string, 0, lcd.rows)
+	if ']' == lcd.nextByteNonWhitespace() {
+		return rows, nil // special case of an empty chunk
+	}
+	lcd.rewind(1)
+
+OuterLoop:
+	for {
+		row, err := lcd.decodeRow()
+		if err != nil {
+			return nil, err
+		}
+		rows = append(rows, row)
+
+		switch c := lcd.nextByteNonWhitespace(); {
+		case c == ',':
+			continue // more elements in the array
+		case c == ']':
+			return rows, nil // we've scanned the whole chunk
+		default:
+			break OuterLoop
+		}
+	}
+	return nil, lcd.mkError("invalid row boundary")
+}
+
+func (lcd *largeChunkDecoder) decodeRow() ([]*string, error) {
+	if '[' != lcd.nextByteNonWhitespace() {
+		return nil, lcd.mkError("expected row to begin with '['")
+	}
+
+	row := make([]*string, 0, lcd.cells)
+	if ']' == lcd.nextByteNonWhitespace() {
+		return row, nil // special case of an empty row
+	}
+	lcd.rewind(1)
+
+OuterLoop:
+	for {
+		cell, err := lcd.decodeCell()
+		if err != nil {
+			return nil, err
+		}
+		row = append(row, cell)
+
+		switch c := lcd.nextByteNonWhitespace(); {
+		case c == ',':
+			continue // more elements in the array
+		case c == ']':
+			return row, nil // we've scanned the whole row
+		default:
+			break OuterLoop
+		}
+	}
+	return nil, lcd.mkError("invalid cell boundary")
+}
+
+func (lcd *largeChunkDecoder) decodeCell() (*string, error) {
+	c := lcd.nextByteNonWhitespace()
+	if c == '"' {
+		s, err := lcd.decodeString()
+		return &s, err
+	} else if c == 'n' {
+		if 'u' == lcd.nextByte() &&
+			'l' == lcd.nextByte() &&
+			'l' == lcd.nextByte() {
+			return nil, nil
+		}
+	}
+	return nil, lcd.mkError("cell begins with unexpected byte")
+}
+
+// TODO we can optimize this further by optimistically searching
+// the read buffer for the next string. If it's short enough and
+// doesn't contain any escaped characters, we can construct the
+// return string directly without writing to the sbuf
+func (lcd *largeChunkDecoder) decodeString() (string, error) {
+	lcd.sbuf.Reset()
+	for {
+		// NOTE if you make changes here, ensure this
+		// variable does not escape to the heap
+		c := lcd.nextByte()
+		if c == '"' {
+			break
+		} else if c == '\\' {
+			if err := lcd.decodeEscaped(); err != nil {
+				return "", err
+			}
+		} else if c < ' ' {
+			return "", lcd.mkError("unexpected control character")
+		} else if c < utf8.RuneSelf {
+			lcd.sbuf.WriteByte(c)
+		} else {
+			lcd.rewind(1)
+			lcd.sbuf.WriteRune(lcd.readRune())
+		}
+	}
+	return lcd.sbuf.String(), nil
+}
+
+func (lcd *largeChunkDecoder) decodeEscaped() error {
+	// NOTE if you make changes here, ensure this
+	// variable does not escape to the heap
+	c := lcd.nextByte()
+
+	switch c {
+	case '"', '\\', '/', '\'':
+		lcd.sbuf.WriteByte(c)
+	case 'b':
+		lcd.sbuf.WriteByte('\b')
+	case 'f':
+		lcd.sbuf.WriteByte('\f')
```

Comment:
what is this for? Looks like no \f is defined in http://www.json.org/.

Reply:
Hm, I was basing this off of the `golang` implementation. 

https://golang.org/src/encoding/json/decode.go (line 1269)